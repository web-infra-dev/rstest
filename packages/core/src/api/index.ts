import { createRunner } from '../runner';
import type {
  Rstest,
  RstestExpect,
  RunnerHooks,
  TestCase,
  TestSummaryResult,
  WorkerState,
} from '../types';
import { GLOBAL_EXPECT, createExpect } from './expect';

export const createRstestRuntime = (
  workerState: WorkerState,
): {
  runner: {
    runTest: (
      testPath: string,
      hooks: RunnerHooks,
    ) => Promise<TestSummaryResult>;
    getCurrentTest: () => TestCase | undefined;
  };
  api: Rstest;
} => {
  const { runner, api: runnerAPI } = createRunner({ workerState });

  const expect: RstestExpect = createExpect({
    workerState,
    getCurrentTest: () => runner.getCurrentTest(),
  });

  Object.defineProperty(globalThis, GLOBAL_EXPECT, {
    value: expect,
    writable: true,
    configurable: true,
  });

  return {
    runner,
    api: {
      expect,
      ...runnerAPI,
    },
  };
};
