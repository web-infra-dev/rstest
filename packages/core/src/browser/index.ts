import type { Rstest } from '../core/rstest';
import { runBrowserController } from './hostController';

export async function runBrowserTests(context: Rstest): Promise<void> {
  await runBrowserController(context);
}
