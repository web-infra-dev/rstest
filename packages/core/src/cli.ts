export { runCLI } from './cli/index';
