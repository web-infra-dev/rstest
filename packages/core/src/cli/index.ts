import { logger } from '../utils/logger';
import { setupCommands } from './commands';
import { prepareCli } from './prepare';

export async function runCLI(): Promise<void> {
  prepareCli();

  try {
    setupCommands();
  } catch (err) {
    logger.error('Failed to start Rstest CLI.');
    logger.error(err);
  }
}
