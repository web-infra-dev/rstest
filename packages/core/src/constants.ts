import type { Rstest } from './types';

export const DEFAULT_CONFIG_NAME = 'rstest.config';

export const DEFAULT_CONFIG_EXTENSIONS = [
  '.js',
  '.ts',
  '.mjs',
  '.mts',
  '.cjs',
  '.cts',
] as const;

export const globalApis: (keyof Rstest)[] = [
  'test',
  'describe',
  'it',
  'expect',
];
