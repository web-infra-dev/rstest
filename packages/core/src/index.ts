export { defineConfig } from './config';

export * from './api/public';

// TODO: Move below to @rstest/node exports point.
export { createRstest } from './core';
export type { RstestConfig } from './types';
