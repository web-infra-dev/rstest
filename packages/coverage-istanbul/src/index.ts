export { pluginCoverage } from './plugin';
export { CoverageProvider } from './provider';
