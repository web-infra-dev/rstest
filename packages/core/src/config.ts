import fs from 'node:fs';
import { dirname, isAbsolute, join } from 'node:path';
import {
  type LoadConfigOptions,
  loadConfig as loadRsbuildConfig,
  mergeRsbuildConfig,
} from '@rsbuild/core';
import { DEFAULT_CONFIG_EXTENSIONS, DEFAULT_CONFIG_NAME } from './constants';
import type { NormalizedConfig, RstestConfig } from './types';
import { color } from './utils/helper';
import { logger } from './utils/logger';

export type RstestConfigAsyncFn = () => Promise<RstestConfig>;

export type RstestConfigSyncFn = () => RstestConfig;

export type RstestConfigExport =
  | RstestConfig
  | RstestConfigSyncFn
  | RstestConfigAsyncFn;

/**
 * This function helps you to autocomplete configuration types.
 * It accepts a Rsbuild config object, or a function that returns a config.
 */
export function defineConfig(config: RstestConfig): RstestConfig;
export function defineConfig(config: RstestConfigSyncFn): RstestConfigSyncFn;
export function defineConfig(config: RstestConfigAsyncFn): RstestConfigAsyncFn;
export function defineConfig(config: RstestConfigExport): RstestConfigExport;
export function defineConfig(config: RstestConfigExport) {
  return config;
}

const findConfig = (basePath: string): string | undefined => {
  return DEFAULT_CONFIG_EXTENSIONS.map((ext) => basePath + ext).find(
    fs.existsSync,
  );
};

const resolveConfigPath = (root: string, customConfig?: string) => {
  if (customConfig) {
    const customConfigPath = isAbsolute(customConfig)
      ? customConfig
      : join(root, customConfig);
    if (fs.existsSync(customConfigPath)) {
      return customConfigPath;
    }
    logger.warn(`Cannot find config file: ${color.dim(customConfigPath)}\n`);
  }

  const configFilePath = findConfig(join(root, DEFAULT_CONFIG_NAME));

  if (configFilePath) {
    return configFilePath;
  }

  return null;
};

export async function loadConfig({
  cwd = process.cwd(),
  path,
  envMode,
  configLoader,
}: {
  cwd?: string;
  path?: string;
  envMode?: string;
  configLoader?: LoadConfigOptions['loader'];
}): Promise<{
  content: RstestConfig;
  filePath: string | null;
}> {
  const configFilePath = resolveConfigPath(cwd, path);

  if (!configFilePath) {
    logger.debug('no config file found');
    return {
      content: {},
      filePath: configFilePath,
    };
  }

  const { content } = await loadRsbuildConfig({
    cwd: dirname(configFilePath),
    path: configFilePath,
    envMode,
    loader: configLoader,
  });

  return { content: content as RstestConfig, filePath: configFilePath };
}

export const mergeRstestConfig = (...configs: RstestConfig[]): RstestConfig =>
  mergeRsbuildConfig<RstestConfig>(...configs);

const createDefaultConfig = (): NormalizedConfig => ({
  root: process.cwd(),
  name: 'rstest',
  include: ['**/*.{test,spec}.?(c|m)[jt]s?(x)'],
  exclude: ['**/node_modules/**', '**/dist/**'],
});

export const withDefaultConfig = (config: RstestConfig): NormalizedConfig => {
  const merged = mergeRstestConfig(createDefaultConfig(), config);

  // The following configurations need overrides
  merged.include = config.include || merged.include;
  merged.exclude = config.exclude || merged.exclude;

  return merged as NormalizedConfig;
};
