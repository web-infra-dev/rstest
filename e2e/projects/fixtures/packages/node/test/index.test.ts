import { sayHi } from '@/src';

it('should test source code correctly', () => {
  expect(sayHi()).toBe('hi');
});

it('should can not get document', () => {
  // @ts-expect-error
  expect(global.document).toBeUndefined();
});

it('should generate snapshot correctly', () => {
  expect('hello world').toMatchSnapshot();
});
