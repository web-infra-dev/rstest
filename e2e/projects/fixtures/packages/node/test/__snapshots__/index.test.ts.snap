// Rstest Snapshot v1

exports[`should generate snapshot correctly 1`] = `"hello world"`;
